### ISSUE TITLE
Duplicate key errors inserting sub-recipients

### ISSUE THREAD

**Issue by ajhyndman** [usdigitalresponse/arpa-reporter#337]
```
revalidate Error: {"error":"update \"uploads\" set \"validated_at\" = $1, \"validated_by\" = $2 where \"id\" = $3 returning * - current transaction is aborted, commands ignored until end of transaction block"}
```

This may also have been resolved with recent changes, but we should re-check after #282 lands.

**Comment by igor47** [comment]
the issue i'm seeing in dev:
```
arpa-reporter-postgres-1  | 2022-06-21 09:37:51.881 UTC [72296] ERROR:  duplicate key value violates unique constraint "arpa_subrecipients_tenant_id_tin_unique"
arpa-reporter-postgres-1  | 2022-06-21 09:37:51.881 UTC [72296] DETAIL:  Key (tenant_id, tin)=(0, 04-2942913) already exists.
arpa-reporter-postgres-1  | 2022-06-21 09:37:51.881 UTC [72296] STATEMENT:  insert into "arpa_subrecipients" ("record", "tin", "uei", "upload_id") values ($1, $2, $3, $4) returning *
arpa-reporter-postgres-1  | 2022-06-21 09:37:52.106 UTC [72296] ERROR:  current transaction is aborted, commands ignored until end of transaction block
```

**Comment by mattbroussard** [comment]
`arpa_subrecipients` has a UNIQUE constraint on `['tenant_id', 'tin']`, but it looks like the code [here](https://github.com/usdigitalresponse/arpa-reporter/blob/1728a883e8af4dced5a75d0d2b38610d67b2b047/src/server/services/validate-upload.js#L148-L158) could do an insert violating this if the same file (or a different file referencing the same TIN / `EIN__c`) is uploaded multiple times, since we only go down the "update existing recipient" path if `existing?.upload_id === upload.id`.

Could this be related to #338?

**Comment by igor47** [comment]
that path is in the [`else` case](https://github.com/usdigitalresponse/arpa-reporter/blob/1728a883e8af4dced5a75d0d2b38610d67b2b047/src/server/services/validate-upload.js#L132) -- prior to that, we attempt to look up the subrecipient [here](https://github.com/usdigitalresponse/arpa-reporter/blob/1728a883e8af4dced5a75d0d2b38610d67b2b047/src/server/services/validate-upload.js#L103) and then if we do find an existing subrecipient, we merely check that the new upload and the existing row match in the main [if clause](https://github.com/usdigitalresponse/arpa-reporter/blob/1728a883e8af4dced5a75d0d2b38610d67b2b047/src/server/services/validate-upload.js#L116)

so, it's kinda odd, but the logic there is pretty convoluted also, i can maybe try to clear it up by pulling out some of the constraints/decision points and giving them names

**Comment by mattbroussard** [comment]
Ah, ok, I didn't notice that this is in an else of a different `existing.upload_id !== upload.id` check -- I think that would mean the `existing?.upload_id === upload.id` check around `updateRecipient` is actually just equivalent to `if (existing)`? Is it possible then `findRecipient` is not correctly finding an existing row that causes the constraint violation (though this seems unlikely to me... `findRecipient` is very straightforward)?

**Comment by igor47** [comment]
closed by #380
